### DISCUSSION TITLE
How to reboot Spike Prime when hung?

### DISCUSSION THREAD

**Discussion by dmusican** [pybricks/support#1165] [Q&A] [answered]
I've been using Pybricks this week with the [Building Block Robotics IR Seeker](https://irseeker.buildingblockrobotics.com/). They've got a description on their website [regarding how to use the sensor with Pybricks](https://irseeker.buildingblockrobotics.com/guides/spike-prime/pybricks). It all generally works, but occasionally when reading from the sensor the hub completely hangs. The LCD screen will show a swirling block as usual while the program is running, but it the code stops running, the hub becomes completely unresponsive to the start/stop button, and the only way we have found to reset it is pop the battery out and back in.

This may be an issue with the sensor itself or its firmware, of course, and I've reached out to the great folks at Building Block Robotics looking for ideas. I'm well aware that resolving issues with that specific sensor is way beyond the scope of this forum. 

But it raises a more fundamental Pybricks question: in the event that the hub hangs, as it does in this case, is there any way I can reset it short of popping the battery? That does the trick, but is quite cumbersome when students have built the hub deeply into a larger project. Are there any other workarounds?

I was previously on the NXT platform, and it had a lovely hidden button deep in one of the LEGO holes that you could reach with a paperclip to do a reset. I've found no such hidden trick for the Spike Prime. It would be great to learn if Pybricks has some other incantation or ritual I can perform to do a reset when hung.

Thanks for the help!

**Comment by dlech** [comment, marked as answer]
It should always be possible to press and hold the power button to turn it off. 

If you have a program that can reproduce the bad behavior, please open an issue and paste the code there along with the firmware version you are using and we will have a look at it. Also please test with the beta version of the firmware to be sure it hasn't been fixed already.

**Comment by dmusican** [comment]
Upgrading to beta solved the problem of the hub locking up (as far as I can reliably test), so thank you!

More details: getting the IR sensor to lock up when [using it as directed](https://irseeker.buildingblockrobotics.com/guides/spike-prime/pybricks) is still something I can't duplicate on demand. It happens rarely and sporadically. But I can write unsupported buggy code that hangs the IR sensor. (Specifically, in the linked code, I change `read(5)` to `read(4)`, which is unsupported and undoubtedly a bad idea). When I do that with the release version of Pybricks, the hub reliably locks up and requires a battery pop, but with the beta Pybricks release I can reset the hub with the power button.

Hopefully, this also works if/when the sensor locks up as it occasionally does when used as directed; we'll only know after trying it for a few hours. I'm optimistic that it will work, as it seems likely it's the same sensor-lockup cause.

**Reply by laurensvalk** [reply]
We've had several support tickets for the Building Block Robotics IR Seeker.

Part of this may be caused by the sensor not sticking to the original LEGO sensor protocol. From the limited information I've seen, it doesn't always return a consistent amount of values when in a particular mode.

It would be helpful if Building Block Robotics reached out to us directly so we could potentially help them fix it.

**Comment by Potato-eater** [comment]
In my experience, firmware v3.3.ob5 works the best with the irseeker. I have had the same issue in the newest firmware versions, but I haven't had any issues when I am using v3.3.0b5.
[pybricks-primehub-v3.3.0b5.zip](https://github.com/pybricks/support/files/12101046/pybricks-primehub-v3.3.0b5.zip)

**Reply by dmusican** [reply]
@Potato-eater: thanks for the suggestion. If we continue to have issues, we'll try out that older firmware version. My students using the IR Seeker worked mostly on physical issues yesterday, so they haven't yet done a good stress test since we updated to the current beta firmware.
